### ISSUE TITLE
Invalid FFI::AutoPointer is causing infinite loop in Finalizer thread starting from JRuby 9.3.4

### ISSUE THREAD

**Issue by rsim** [appsignal/appsignal-ruby#854]
After upgrading from JRuby 9.3.3.0 to 9.3.6.0 we noticed that our Rails app with the appsignal gem consumes more CPU. From the thread dump we identified that the Java Finalizer thread is stuck in an infinite loop:

```
"Finalizer" - Thread t@3
   java.lang.Thread.State: RUNNABLE
        at org.jruby.RubyException.checkCircularCause(RubyException.java:399)
        at org.jruby.RubyException.setCause(RubyException.java:370)
        at org.jruby.Ruby.newRaiseException(Ruby.java:4299)
        at org.jruby.Ruby.newArgumentError(Ruby.java:3631)
        at org.jruby.Ruby.newArgumentError(Ruby.java:3626)
        at org.jruby.runtime.Arity.checkArity(Arity.java:167)
        at org.jruby.runtime.Arity.checkArity(Arity.java:161)
        at org.jruby.ext.ffi.jffi.NativeInvoker.call(NativeInvoker.java:70)
        at org.jruby.ext.ffi.jffi.DefaultMethod.call(DefaultMethod.java:106)
        at org.jruby.internal.runtime.methods.DynamicMethod.call(DynamicMethod.java:222)
        at org.jruby.RubyMethod.call(RubyMethod.java:119)
        at org.jruby.RubyMethod$INVOKER$i$call.call(RubyMethod$INVOKER$i$call.gen)
        at org.jruby.internal.runtime.methods.JavaMethod$JavaMethodZeroOrOneOrNBlock.call(JavaMethod.java:354)
        at org.jruby.runtime.callsite.CachingCallSite.call(CachingCallSite.java:173)
        at home.rails.eazybi.production.shared.bundle.jruby.$2_dot_6_dot_0.gems.ffi_minus_1_dot_15_dot_5_minus_java.lib.ffi.autopointer.invokeOther1:call(/home/rails/eazybi/production/shared/bundle/jruby/2.6.0/gems/ffi-1.15.5-java/lib/ffi/autopointer.rb:175)
        at home.rails.eazybi.production.shared.bundle.jruby.$2_dot_6_dot_0.gems.ffi_minus_1_dot_15_dot_5_minus_java.lib.ffi.autopointer.RUBY$method$release$0(/home/rails/eazybi/production/shared/bundle/jruby/2.6.0/gems/ffi-1.15.5-java/lib/ffi/autopointer.rb:175)
        at java.lang.invoke.LambdaForm$DMH/0x0000000801af7840.invokeStatic(LambdaForm$DMH)
        at java.lang.invoke.LambdaForm$MH/0x00000008006f2040.invokeExact_MT(LambdaForm$MH)
        at org.jruby.internal.runtime.methods.CompiledIRMethod.call(CompiledIRMethod.java:165)
        at org.jruby.internal.runtime.methods.MixedModeIRMethod.call(MixedModeIRMethod.java:185)
        at org.jruby.internal.runtime.methods.DynamicMethod.call(DynamicMethod.java:218)
        at org.jruby.runtime.callsite.CachingCallSite.call(CachingCallSite.java:173)
        at home.rails.eazybi.production.shared.bundle.jruby.$2_dot_6_dot_0.gems.ffi_minus_1_dot_15_dot_5_minus_java.lib.ffi.autopointer.invokeOther3:release(/home/rails/eazybi/production/shared/bundle/jruby/2.6.0/gems/ffi-1.15.5-java/lib/ffi/autopointer.rb:150)
        at home.rails.eazybi.production.shared.bundle.jruby.$2_dot_6_dot_0.gems.ffi_minus_1_dot_15_dot_5_minus_java.lib.ffi.autopointer.RUBY$method$call$0(/home/rails/eazybi/production/shared/bundle/jruby/2.6.0/gems/ffi-1.15.5-java/lib/ffi/autopointer.rb:150)
        at java.lang.invoke.LambdaForm$DMH/0x0000000801af7440.invokeStatic(LambdaForm$DMH)
        at java.lang.invoke.LambdaForm$MH/0x00000008006f2040.invokeExact_MT(LambdaForm$MH)
        at org.jruby.internal.runtime.methods.CompiledIRMethod.call(CompiledIRMethod.java:139)
        at org.jruby.internal.runtime.methods.CompiledIRMethod.call(CompiledIRMethod.java:162)
        at org.jruby.internal.runtime.methods.MixedModeIRMethod.call(MixedModeIRMethod.java:185)
        at org.jruby.internal.runtime.methods.DynamicMethod.call(DynamicMethod.java:218)
        at org.jruby.runtime.callsite.CachingCallSite.call(CachingCallSite.java:173)
        at org.jruby.RubyBasicObject$Finalizer.callFinalizer(RubyBasicObject.java:1971)
        at org.jruby.RubyBasicObject$Finalizer.finalize(RubyBasicObject.java:1960)
        at java.lang.System$2.invokeFinalize(System.java:2125)
        at java.lang.ref.Finalizer.runFinalizer(Finalizer.java:87)
        at java.lang.ref.Finalizer$FinalizerThread.run(Finalizer.java:171)
```

Infinite loop is caused by `checkCircularCause` which was added in https://github.com/jruby/jruby/commit/bd2595c275adb0a03df8f281b6b144718f06e85d which solved https://github.com/jruby/jruby/issues/7035. It seems that somehow circular exception cause loop is created which does not include the latest top exception.

But further in the stack we can see that the root cause is `FFI::AutoPointer` at `autopointer.rb:175` which calls
```ruby
@proc.call(ptr)
```
and then `checkArity` fails for this call.

As I found the `appsignal` gem defines `FFI::AutoPointer` in three places:
```ruby
@pointer = FFI::AutoPointer.new(
  pointer,
  Extension.method(:appsignal_free_transaction)
)

@pointer = FFI::AutoPointer.new(
  pointer,
  Extension.method(:appsignal_free_span)
)

@pointer = FFI::AutoPointer.new(
  pointer,
  Extension.method(:appsignal_free_data)
)
```

And these methods are defined as foreign functions in the following way:
```ruby
attach_function :appsignal_free_transaction,
  [],
  :void

attach_function :appsignal_free_span,
  [:pointer],
  :void

attach_function :appsignal_free_data, [], :void
```

As I see `appsignal_free_span` is defined with a `:pointer` argument but `appsignal_free_transaction` and `appsignal_free_data` are defined without any arguments which I suspect is wrong.

My guess is that when `FFI::AutoPointer` tries to call the finalizer code with either `appsignal_free_transaction` or `appsignal_free_data` then the method arity check fails (as it tries to call it with a pointer argument but they are defined without arguments). And now the new `checkCircularCause` method results in an infinite loop and exposes this problem, previously I suspect it was silently failing and was not noticeable.

**Comment by shairyar** [comment]
https://app.intercom.com/a/apps/yzor8gyw/inbox/inbox/4356044/conversations/16410700130776 (Private link)

**Comment by rsim** [comment]
We patched the appsignal gem with this change https://github.com/eazybi/appsignal-ruby/commit/bae84fd176c473a5812d44a314157b09633ab2be and now in production, we do not see anymore the Finalizer thread with an infinite loop in `checkCircularCause`. Please verify this patch and include in the next version of the gem.

**Comment by shairyar** [comment]
Thanks @rsim for the patch, we will take a look into that.

**Comment by tombruijn** [comment]
I've created a PR for the fix you sent in here: https://github.com/appsignal/appsignal-ruby/pull/856

**Comment by tombruijn** [comment]
Looks like our CI didn't show us this issue because we don't test against JRuby 9.3, in which [the commit that highlights this issue](https://github.com/jruby/jruby/commit/bd2595c275adb0a03df8f281b6b144718f06e85d) was added. I've created an issue to add JRuby 9.3 to our CI build in #858.

**Comment by tombruijn** [comment]
We're hoping to release the fix for this next week.

**Comment by tombruijn** [comment]
@rsim The fix for this has been released in Ruby gem 3.1.0. And we're already up to 3.1.2 now :)
